nix/docker/context/bin/run-node

```diff
@@ -0,0 +1,276 @@
+CNODE_HOSTNAME="$CARDANO_PUBLIC_IP"
+CNODE_PORT=$CARDANO_PORT
+CUSTOM_PEERS="$CARDANO_CUSTOM_PEERS"
+GENESIS_JSON="$CARDANO_CONFIG_BASE/mainnet-shelley-genesis.json"
+TOPOLOGY="$CARDANO_TOPOLOGY"
+LOG_DIR="$CARDANO_LOG_DIR"
+EKG_HOST="$EKG_HOST"
+EKG_PORT=$EKG_PORT
+EKG_TIMEOUT=$EKG_TIMEOUT
+MAX_PEERS=$MAX_PEERS
+
+# Mapping for gLiveView
+CCLI="cardano-cli"
+CNODE_PORT=$CARDANO_PORT
+CONFIG="$CARDANO_CONFIG"
+TOPOLOGY="$CARDANO_TOPOLOGY"
+SOCKET="$CARDANO_SOCKET_PATH"
+LOG_DIR="$CARDANO_LOG_DIR"
+EOF
+}
+
+#####################################################################
+#
+# Run the relay node in the background
+#
+runRelayNode () {
+
+  effopts=(--config $CARDANO_CONFIG \
+    --topology $CARDANO_TOPOLOGY \
+    --database-path $CARDANO_DATABASE_PATH \
+    --socket-path $CARDANO_SOCKET_PATH \
+    --host-addr $CARDANO_BIND_ADDR \
+    --port $CARDANO_PORT)
+
+  effopts+=(${options[@]})
+
+  echo "cardano-node run ${effopts[@]}"
+  cardano-node run ${effopts[@]} &
+}
+
+#####################################################################
+#
+# Run the block producer in the background
+#
+runBlockProducerNode () {
+
+  effopts=(--config $CARDANO_CONFIG \
+    --topology $CARDANO_TOPOLOGY \
+    --database-path $CARDANO_DATABASE_PATH \
+    --socket-path $CARDANO_SOCKET_PATH \
+    --host-addr $CARDANO_BIND_ADDR \
+    --port $CARDANO_PORT \
+    --shelley-kes-key $CARDANO_SHELLEY_KES_KEY \
+    --shelley-vrf-key $CARDANO_SHELLEY_VRF_KEY \
+    --shelley-operational-certificate $CARDANO_SHELLEY_OPERATIONAL_CERTIFICATE)
+
+  effopts+=(${options[@]})
+
+  echo "cardano-node run ${effopts[@]}"
+  cardano-node run ${effopts[@]} &
+}
+
+# Override default values with explicit options
+options=($@)
+for i in "${!options[@]}"
+do
+    j=$((i + 1))
+    key=${options[i]}
+    val=${options[j]}
+    found=false
+
+    # echo "$i/$j: ${key} ${val}"
+
+    case ${key} in
+      --config) CARDANO_CONFIG=${val}; found=true;;
+      --topology) CARDANO_TOPOLOGY=${val}; found=true;;
+      --database-path) CARDANO_DATABASE_PATH=${val}; found=true;;
+      --socket-path) CARDANO_SOCKET_PATH=${val}; found=true;;
+      --host-addr) CARDANO_BIND_ADDR=${val}; found=true;;
+      --port) CARDANO_PORT=${val}; found=true;;
+      --shelley-kes-key) CARDANO_SHELLEY_KES_KEY=${val}; found=true;;
+      --shelley-vrf-key) CARDANO_SHELLEY_VRF_KEY=${val}; found=true;;
+      --shelley-operational-certificate) CARDANO_SHELLEY_OPERATIONAL_CERTIFICATE=${val}; found=true;;
+    esac
+
+    if [[ $found == true ]]; then
+      options[i]="";
+      options[j]="";
+    fi
+done
+
+printRunEnv
+writeRootEnv
+
+# The IPC socket dir is not created on demand
+mkdir -p `dirname $CARDANO_SOCKET_PATH`
+
+# Removing a potentialy existing lost+found directory
+rm -rf $CARDANO_DATABASE_PATH/lost+found
+
+if [[ $CARDANO_BLOCK_PRODUCER == true ]]; then
+
+  runBlockProducerNode
+
+else
+
+  if [[ $CARDANO_UPDATE_TOPOLOGY == true ]]; then
+    topologyUpdate -l &
```

Comment:
so the topology update is done every hour in the background. But currently the node need to be restarted to pickup any change to the topology file (since will change with p2p support).

Should a synchronous update be done instead (or in addition to)?

Reply:
Yes, we run the push/fetch for the topology update once per hour in the background. We used to do this in a cron job that ran inside the docker container (just like almost every other node). This however created an unnecessary dependency on base image specifics (i.e. on the OS) which unfortunately are not the same across supported architectures (i.e. its different for arm64).

How does a non-docker node pick up changes to the topology file when they are done async from a cron job? To be honest, I never questioned this because I assumed that the node somehow "watches" its topology updates. The async cron job approach suggested this and the the above has always worked (for many months so far).

Reply:
Currently we restart the node every time the topology is updated.

Reply:
Who restarts the node, is the cron job supposed to do that? A restart is not instantaneous, especially not if graceful shutdown is not done properly. Even when it is, it may take minutes for the node to communicate with the BP again, which may lead to block loss. 

Perhaps it is sufficient to publish the node's IP/DNS to be connected well and it is not actually necessary to read the updated topology. Restarting the node every hour seems quite wrong and I somehow can't believe that many SPOs actually do that.

Reply:
We could also remove the topology update from this PR. It isn't part of the existing image and would be taken out with Alonzo anyway.